Add user notification API

eviljeff opened the pull request:
fixes #5408

eviljeff commented:
Not sure about the API docs - in both the list and set (update) cases it's not a single JSON object, it's either a list or a dict (which I guess _is_ a JSON object, just one without pre-defined properties).  The magic syntax sphinx uses doesn't seem to support that.

diox commented on src/olympia/accounts/views.py:
```diff
@@ -487,3 +490,56 @@ def post(self, request):
+                self._get_default_object(notification)))  # Otherwise, default.
+        return out
+
+    @list_route(methods=['patch'])
```
The `method` is the only thing I'm not sure about. Since it's a dedicated endpoint, I would have expected a `post` instead. Am I the only one ?

EnTeQuAk commented on docs/topics/api/accounts.rst:
```diff
@@ -187,6 +187,46 @@ This endpoint lists the add-ons in a collection, together with collector's notes
+    permission to list notifications on accounts other than your own.
+
+This endpoint allows you to list the account notifications set for the specified user.
+The result is an unpaginated list of the fields below.  There are currently 11 notifications.
```
Double spaces after the dot

EnTeQuAk replied:
"There are currently 11 notification types" maybe?

EnTeQuAk commented on docs/topics/api/accounts.rst:
```diff
@@ -187,6 +187,46 @@ This endpoint lists the add-ons in a collection, together with collector's notes
+
+    :>json string name: notification short name
+    :>json boolean enabled: if the notification is enabled (defaults to True)
+    :>json boolean mandatory: if the notification can be set by the user.
```
Usually these descriptions are full sentences, meaning: Capital start and dot at the end.

EnTeQuAk commented on docs/topics/api/accounts.rst:
```diff
@@ -187,6 +187,46 @@ This endpoint lists the add-ons in a collection, together with collector's notes
+This endpoint allows account notifications to be set. The request should be a dict of `name`:True|False pairs.
+Any number of notifications can be set; only non-mandatory notifications can be changed - attempting to set a mandatory notification will be ignored.
+
+.. http:patch:: /api/v3/accounts/account/(int:user_id|string:username)/notifications/set
```
I actually wonder why this is a separate endpoint? Can't this be handled in the `/notifications/` endpoint itself? The separation that this is "setting some value" should be handled by specifying the http verb `PATCH` already.

kumar303 replied:
I agree that it's a little confusing to see this as a separate endpoint

EnTeQuAk commented on src/olympia/accounts/views.py:
```diff
@@ -487,3 +490,56 @@ def post(self, request):
+
+
+class AccountNotificationViewSet(ListModelMixin, GenericViewSet):
+    """Returns account notifications.  If not already set by the user defaults
```
More double whitespaces, also the style is a bit wrong, usually it's something along the lines of

```python
"""Returns account notifications.

If not already set by the user, defaults will be returned.
"""
```

According to pep08 and other style guides (imho)

kumar303 commented on docs/topics/api/accounts.rst:
```diff
@@ -187,6 +187,46 @@ This endpoint lists the add-ons in a collection, together with collector's notes
+    This API requires :doc:`authentication <auth>` and `Users:Edit`
+    permission to set notification preferences on accounts other than your own.
+
+This endpoint allows account notifications to be set. The request should be a dict of `name`:True|False pairs.
```
It would be great if one day the docs had a canonical section on how you can define boolean values in DRF.

diox replied:
well, 99% of the case (here is no exception) we'll be dealing with json :)

kumar303 commented on docs/topics/api/accounts.rst:
```diff
@@ -187,6 +187,46 @@ This endpoint lists the add-ons in a collection, together with collector's notes
+    permission to set notification preferences on accounts other than your own.
+
+This endpoint allows account notifications to be set. The request should be a dict of `name`:True|False pairs.
+Any number of notifications can be set; only non-mandatory notifications can be changed - attempting to set a mandatory notification will be ignored.
```
> attempting to set a mandatory notification will be ignored.

This is going make someone lose a lot of hair one day :) Can it return a [409](https://httpstatuses.com/409) instead of ignoring a request to change a mandatory notification?

eviljeff replied:
Standard behaviour for trying to set a read-only field on a serializer is for it to be silently ignored so it seemed appropriate to replicate.

eviljeff replied:
changed in https://github.com/mozilla/addons-server/pull/5435/commits/db886fb2e25dc9787f1d8679440c9ce4497a92db

kumar303 replied:
Thanks for the change. This will make using the API much nicer.

eviljeff commented:
Because implementing PATCH on the standard list endpoint would be messy (requiring subclassing SimpleRouter and modifying self.routes); instead changed it to POST for easy maintenance.

Updated docs with suggestions.

Changed so attempting to set a mandatory notification throws an exception (400)
